Refresh button for source rows in Status page

lovincyrus opened the pull request:
address https://github.com/rilldata/rill/issues/6154

- [x] invalidate resources data when refresh source is clicked
- [x] unable to invalidate individual source resource

![CleanShot 2024-11-26 at 13 44 08@2x](https://github.com/user-attachments/assets/fd296843-ac42-463f-8fed-0cb41d85ba6a)

ericpgreen2 commented:
Note: we should also include the refresh button for Models because 1) a Model can be a root node of the DAG, 2) an intermediate model could still have logic that needs a refresh independent from an upstream node, and 3) we're soon deprecating the Source concept anyways.

lovincyrus commented:
> Note: we should also include the refresh button for Models because 1) a Model can be a root node of the DAG, 2) an intermediate model could still have logic that needs a refresh independent from an upstream node, and 3) we're soon deprecating the Source concept anyways.
> 
> And actually, if it's not too much a scope expansion, it'd be nice to be able to refresh Alerts & Reports too. See: <img alt="image" width="636" src="https://private-user-images.githubusercontent.com/14206386/390474766-5ea84128-3243-427a-9e09-802ae3eb53a4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMyNzIwNjcsIm5iZiI6MTczMzI3MTc2NywicGF0aCI6Ii8xNDIwNjM4Ni8zOTA0NzQ3NjYtNWVhODQxMjgtMzI0My00MjdhLTllMDktODAyYWUzZWI1M2E0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA0VDAwMjI0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI2MmNkYjExYjU0MGQyMTcwZWFkZGRjZjQwZTFkZjA2MzNhZWIyODc5MzE0YjJmNDZjNjNiNGEwYTBiYjU1YjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.L3HY5hBBDdPyCWerR7x0tZzn3GR96TzCUVQHPYxb87Q">

Added row click refresh button to `Source` and `Model` resource kind. I also added a tooltip to indicate that a resource trigger will update all dependent resources.

---

Latest demo:

https://github.com/user-attachments/assets/393afa07-1076-4792-9d3b-ab99274609ec

ericpgreen2 reviewed (changes requested):
There's a lot of action when clicking the refresh button: flickering, jumpiness, table resizing. Some things I noticed:
1. The Resources table unmounts temporarily. [See Jam](https://jam.dev/c/941ace36-aa0c-4e77-afc4-27c89cf3ecd1) and screenshot:
    <img width="1426" alt="image" src="https://github.com/user-attachments/assets/f6e2ae6c-49d3-4a5c-ad34-cb292c55ef98">
2. The page gets auto-scrolled to the top ([Jam](https://jam.dev/c/941ace36-aa0c-4e77-afc4-27c89cf3ecd1)), probably due to point 1.
3. A `RefreshTrigger` resource is shown temporarily. Can we hide it from the list? Here's a screenshot from your video:
    <img width="1664" alt="image" src="https://github.com/user-attachments/assets/7e229e35-7280-446d-ab65-d70f7d78b6ee">

lovincyrus commented:
Latest demo:

https://github.com/user-attachments/assets/209550f4-9cbd-47f1-b27c-6c361779a766

ericpgreen2 commented on scripts/tsc-with-whitelist.sh:
```diff
@@ -63,6 +63,7 @@ web-common/src/runtime-client/http-request-queue/HttpRequestQueue.ts: error TS25
 web-common/src/runtime-client/http-request-queue/HttpRequestQueueTypes.ts: error TS18048
 web-common/src/runtime-client/http-request-queue/HttpRequestQueueTypes.ts: error TS2322
 web-common/src/lib/time/config.ts: error TS2322
+web-common/src/features/file-explorer/new-files.ts: error TS2741
```
Is this a challenging type error? If easy, it'd be good to fix rather than expand this list.

lovincyrus replied:
I'm not sure if I made any changes that caused this in this PR

ericpgreen2 replied:
I think it's from this addition to the `ResourceKind` enum:

```
RefreshTrigger = "rill.runtime.v1.RefreshTrigger",
```

ericpgreen2 reviewed (commented):
Mostly looks good, but I ran into a weird state: ~50% of the time when refreshing a Source with an error, the source is marked "Idle", yet for about 30s the "Refresh all sources and models" button is disabled and I see polling `ListResources` network requests. [Here's a Jam](https://jam.dev/c/6ca87666-2f41-412e-adb2-5996017189d9).

I also hit this 500 error, but I think it's been fixed on `main`, right? So it'd help if you could merge `main` into this branch.
<img width="852" alt="image" src="https://github.com/user-attachments/assets/4ea711a2-a2fb-454d-9372-76327550cb70">

lovincyrus commented:
> Mostly looks good, but I ran into a weird state: ~50% of the time when refreshing a Source with an error, the source is marked "Idle", yet for about 30s the "Refresh all sources and models" button is disabled and I see polling `ListResources` network requests. [Here's a Jam](https://jam.dev/c/6ca87666-2f41-412e-adb2-5996017189d9).
> 
> I also hit this 500 error, but I think it's been fixed on `main`, right? So it'd help if you could merge `main` into this branch. <img alt="image" width="852" src="https://private-user-images.githubusercontent.com/14206386/394023138-4ea711a2-a2fb-454d-9372-76327550cb70.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM3OTc4OTcsIm5iZiI6MTczMzc5NzU5NywicGF0aCI6Ii8xNDIwNjM4Ni8zOTQwMjMxMzgtNGVhNzExYTItYTJmYi00NTRkLTkzNzItNzYzMjc1NTBjYjcwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjEwVDAyMjYzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBmMmU2NmNhYjIxZTMyZTJkY2UzODQxZWQ3YjQzMWRiMTYzNWM4Y2ExYzA5MzA0M2I0OTg3OTRmYmFlNTExNDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.i6-iG-acX9Hkji0sPfdGYagm5cXC385hktRoXYcpeBA">

I can see from the Jam that the specific resource didn't get refreshed properly due to the "failed to ingest source" error. Here's a video of a happy path:

https://github.com/user-attachments/assets/5a4b5abd-1b97-4635-b1ef-297a2eb383ab

Is it expected to disable the refresh icon when a resource errors out? I can see in the Jam, the resource continues to error out even after a refresh..

lovincyrus commented:
> I also hit this 500 error, but I think it's been fixed on `main`, right? So it'd help if you could merge `main` into this branch.
> <img alt="image" width="852" src="https://private-user-images.githubusercontent.com/14206386/394023138-4ea711a2-a2fb-454d-9372-76327550cb70.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM4MDY0NTEsIm5iZiI6MTczMzgwNjE1MSwicGF0aCI6Ii8xNDIwNjM4Ni8zOTQwMjMxMzgtNGVhNzExYTItYTJmYi00NTRkLTkzNzItNzYzMjc1NTBjYjcwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjEwVDA0NDkxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNkOGFkMTdhNmNhZDZmM2JiYzQxM2NmYTE5MjVkNzQ2NzM4MjFjMTBiODMzMDZkOWZlMjNkYmFjM2ZjMDU5YWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.QKfLrC2UjbNpBnhcpEQWZ3T60vp_OxS7cFR04yjtXoQ">

After a rebase, I still run into the 500 error occasionally when I fire a `$createTrigger` call from the individual rows. This doesn't need to block this PR. Would you have more context? @begelundmuller 

Error message — "grpc: error while marshaling: marshaling rill.runtime.v1.ListResourcesResponse: size mismatch (see https://github.com/golang/protobuf/issues/1609): calculated=229, measured=177"

![CleanShot 2024-12-10 at 14 01 21@2x](https://github.com/user-attachments/assets/22eb9fdb-b6ae-4f29-9ec4-c4f40df15c91)

begelundmuller commented:
> After a rebase, I still run into the 500 error occasionally when I fire a `$createTrigger` call from the individual rows. This doesn't need to block this PR. Would you have more context? @begelundmuller

Yeah that should be fixed by this PR: https://github.com/rilldata/rill/pull/6233

ericpgreen2 reviewed (changes requested):
UXQA: I've triggered a Source refresh, yet I don't see any polling of the `ListResource` API.

https://github.com/user-attachments/assets/28609b30-fe85-4934-80f3-de487f4432ae

ericpgreen2 commented on web-admin/src/features/projects/status/ProjectResources.svelte:
```diff
@@ -10,41 +10,104 @@
         refetchOnMount: true,
-        refetchInterval: isReconciling ? 500 : false,
+        keepPreviousData: true,
+        onError: () => {
+          stopPolling();
+        },
```
1. `onError` will be deprecated in TanStack Query v5 ([see here](https://tkdodo.eu/blog/breaking-react-querys-api-on-purpose))
2. Rather than the imperative `startPolling()`, `stopPolling()`, have you considered using the built-in declarative `refetchInterval: () => number | false`?

lovincyrus replied:
Just realized that my response was stuck in a pending state, sorry!

Update: Okay, I tried out refresh interval but the imperative polling seems to be a more straightforward approach in updating the resource status. With the refresh interval approach, the loading spinner would remain stuck even after the resource had been reconciled. It's easier to stop polling when different conditions have been met.

ericpgreen2 replied:
Are you sure about this? This is precisely what a `refetchInterval` function is supposed to be used for. The imperative logic – split over a few functions and a handful of reactive statements – is looking pretty complicated. 

I'm able to break the current approach by: refreshing a source, navigating away from the status page, and navigating back to the status page. I'm observing that the polling has stopped, yet I'd expect that the polling continues.

Ideally (as mentioned below), we can make a `refetchInterval` function that we can re-use for the `ListDashboards`, `ListReports`, and `ListAlerts` requests on those resources' respective pages.

lovincyrus replied:
Referred to https://github.com/rilldata/rill/pull/6421 and it worked

lovincyrus commented:
> UXQA: I've triggered a Source refresh, yet I don't see any polling of the `ListResource` API.
> 
>  Kapture.2024-12-16.at.14.40.05.mp4

I cannot consistently reproduce this..

https://github.com/user-attachments/assets/d104c6c4-c214-468f-b605-ad44f81a2d55

lovincyrus commented:
> UXQA: I've triggered a Source refresh, yet I don't see any polling of the `ListResource` API.
> 
>  Kapture.2024-12-16.at.14.40.05.mp4

In this PR, I've used the `ListResources` API to refresh a resource.

Taking another look at this,
![image](https://github.com/user-attachments/assets/1dfaaa19-fbed-4edf-83d6-cafaa519c016)

Specifically, the Refresh a model command.
![CleanShot 2024-12-18 at 23 53 52@2x](https://github.com/user-attachments/assets/087040c6-1fbf-40c0-a960-171b0afd0a3a)

I see `runtimev1.RefreshModelTrigger` and `runtimev1.CreateTriggerRequest` from the `refresh.go`. Any reason why we don't have `RefreshModelTrigger` runtime service? @ericpgreen2

begelundmuller commented:
@lovincyrus @ericpgreen2 FYI `RefreshModelTrigger` is used in `CreateTriggerRequest.models` ([link](https://github.com/rilldata/rill/blob/221b323aff68072a164c86a1bdb9cfda5c665293/proto/rill/runtime/v1/api.proto#L697)) to enable fine-grained refreshes of models (incremental/full refreshes or refreshes of a specific partition). You can still use specify a model in `CreateTriggerRequest.resources` ([link](https://github.com/rilldata/rill/blob/221b323aff68072a164c86a1bdb9cfda5c665293/proto/rill/runtime/v1/api.proto#L695)), but then it will always be a "normal" model refresh (which is a full refresh for non-incremental models and an incremental refresh for incremental models).

ericokuma commented:
<img width="599" alt="Screenshot 2025-01-17 at 2 59 46 PM" src="https://github.com/user-attachments/assets/acb71038-d56b-471c-b6b9-3ae49445a844" />
I'm getting a 400 error when trying to access this PR

ericokuma commented:
from @jkhwu: After they click refresh we can display a temp toast “🔘 Refreshing <resource name>” Followed by another toast “✔️ Successfully refreshed <resource name>”

ericokuma commented:
Thanks Cyrus! Can we do the following changes:

1. When a user clicks on refresh, surface a confirmation prompt that reads: "Refresh {resource_name}? Two buttons: `Cancel` and `Yes, refresh`. 

2. When user clicks on `Yes, refresh`, it'll trigger a toast notification that reads "Refreshing {resource_name}

3. When refresh successful, it'll trigger a toast notification that reads "Successfully refreshed {resource_name}"

4. When refresh fail, it'll trigger a toast notification that reads "Refreshing {resource_name} failed" Action: `Try again` which reattempts the refresh

cc @jkhwu, what icons would want in the toast notifications?

lovincyrus commented:
> cc @jkhwu, what icons would want in the toast notifications?

I'll reuse the existing icons in the codebase.

ericokuma commented:
@lovincyrus, great work here! Only feedback is for the initial toast notification, can we not show the "x"? In this context, it feels like you can click on the "X" to cancel the refresh

<img width="448" alt="Screenshot 2025-01-27 at 9 38 05 AM" src="https://github.com/user-attachments/assets/15fbda80-bcba-458b-b178-deecdaced907" />

lovincyrus commented:
1, 2. The reason is that there is no promise-based notification implementation. I used the persisted option as a workaround to keep the notification visible while waiting for the next success or error status. The "x" in the notification indicates that the notification is being dismissed, not the inflight request being canceled.

3. I believe this is expected when we attempt to refresh a resource. We get a new refresh timestamp regardless of the status.

ericokuma commented:
> 1, 2. The reason is that there is no promise-based notification implementation. I used the persisted option as a workaround to keep the notification visible while waiting for the next success or error status. The "x" in the notification indicates that the notification is being dismissed, not the inflight request being canceled.
> 
> 3. I believe this is expected when we attempt to refresh a resource. We get a new refresh timestamp regardless of the status.

#1,2: Gotcha. The persisted option for when a success toast notification (without the "x") is displayed might not be the right choice. We should remove the notification after a certain amount of time (8 seconds) if it's a toast notification that doesn't have the "x", otherwise, there's no way to dismiss it. for the "refresing resource_name" notification, perhaps we remove the "x" in that toast then and when it either fails or succeeds, we auto-dismiss it

#3: might be misleading for it to say it has been refreshed when the refresh attempt failed

lovincyrus commented:
> #3: might be misleading for it to say it has been refreshed when the refresh attempt failed

This requires platform support to not update the timestamp when a trigger attempt fails.

ericokuma commented:
> > #3: might be misleading for it to say it has been refreshed when the refresh attempt failed
> 
> This requires platform support to not update the timestamp when a trigger attempt fails.

Gotcha! fyi @begelundmuller

it's probably not a blocker for release

ericokuma commented:
https://www.loom.com/share/968406d6fc8a410d983d83ae106bc725?sid=f0ece7e1-5560-4fad-8620-26ee55bc65c4

For the "Refreshing {resource_name}" notification, it should only dismiss once we've received either a success or failure response

ericokuma commented:
![Screenshot 2025-01-27 at 4 09 04 PM](https://github.com/user-attachments/assets/78056e0b-aeca-4064-9c7a-cfedf08fc119)

im still seeing that the success toast notification persists.

ericokuma commented:
but other than that, the PR is good!

ericokuma commented:
Looks good to me!

ericpgreen2 reviewed (changes requested):
After 30s, I see this "Failed to refresh" message, yet the resources still have a spinner refresh indicator.

Can we not stop polling after 30s? To avoid hitting the runtime with requests twice-a-second for perpetuity, could we use a backoff function that slows polling to once per 10s and continues forever?

Note: we will also want to use this polling logic on the each of the project resources tables (Dashboards/Alerts/Reports).

<img width="1330" alt="image" src="https://github.com/user-attachments/assets/3e623e6a-8146-4e20-96a7-cb4c64c7f02b" />

ericpgreen2 commented on web-admin/src/features/projects/status/ProjectResources.svelte:
```diff
@@ -2,54 +2,154 @@
+    }),
+    refetchInterval: (data) => {
+      if (
+        !isPollingEnabled ||
```
We shouldn't need this `isPollingEnabled` variable. The `refetchInterval` function should dictate whether or not there is polling happening. If the conditions do not merit polling (all resources are idle), the function should return `false`.

Though, I could see you wanting to set an `isCurrentlyPolling` variable inside the `refetchInterval` function, so that you could know its polling status outside the hook.

ericpgreen2 commented on web-admin/src/features/projects/status/ProjectResources.svelte:
```diff
@@ -2,54 +2,154 @@
+    refetchInterval: (data) => {
+      if (
+        !isPollingEnabled ||
+        $resources?.isError ||
```
This looks circular. The `refetchInterval` function should be a pure function that only cares about the `data`.

ericpgreen2 commented on web-admin/src/features/projects/status/ProjectResources.svelte:
```diff
@@ -2,54 +2,154 @@
+      // Calculate time elapsed since polling started
+      const elapsedTime = Date.now() - pollStartTime;
+
+      // After threshold, gradually increase interval to MAX_POLL_INTERVAL
+      if (elapsedTime > BACKOFF_THRESHOLD_MS) {
+        return MAX_POLL_INTERVAL;
+      }
+
+      return INITIAL_POLL_INTERVAL;
```
Nit: can we do more graceful scaling? Firing 30 requests in 30 seconds seems a little heavy.

ericpgreen2 commented on web-common/src/lib/event-bus/events.ts:
```diff
@@ -4,15 +4,16 @@ interface Link {
 export interface NotificationMessage {
-  type?: "default" | "success" | "error";
-  message: string;
+  type?: "default" | "success" | "loading" | "error" | "clear-all";
```
All these types refer to different kinds of toasts. Rather than adding a `clear-all` type, I think it'd be cleaner to add a new `clear-all-notifications` event altogether.

lovincyrus replied:
for sure, it could go one or another, that's why we have a review process here :)

ericpgreen2 commented on web-common/src/features/file-explorer/new-files.ts:
```diff
@@ -116,6 +116,11 @@ export const ResourceKindMap: Record<
     baseName: "alert",
     extension: ".yaml",
   },
+  [ResourceKind.RefreshTrigger]: {
+    folderName: "triggers",
+    baseName: "refresh_trigger",
+    extension: ".yaml",
+  },
```
This list is "user-facing kinds", and I wouldn't consider the refresh trigger user-facing. Instead, you could add `RefreshTrigger` to the exclusion list here:
```
export type UserFacingResourceKinds = Exclude<
  ResourceKind,
  ResourceKind.ProjectParser,
  ResourceKind.RefreshTrigger // Add this
>;
```

lovincyrus replied:
thanks!

lovincyrus commented on web-admin/src/features/projects/status/selectors.ts:
```diff
@@ -18,3 +25,18 @@ export function useProjectDeployment(orgName: string, projName: string) {
   );
 }
+
+export function useResources(
+  instanceId: string,
+  queryOptions?: CreateQueryOptions<
+    V1ListResourcesResponse,
+    ErrorType<RpcStatus>,
+    V1ListResourcesResponse
+  >,
+) {
+  return createRuntimeServiceListResources(instanceId, undefined, {
+    query: {
+      ...queryOptions,
+    },
+  });
+}
```
this seems like a pattern from the other part of the codebase, https://github.com/rilldata/rill/blob/main/web-common/src/features/explores/selectors.ts#L28

lovincyrus replied:
i'll revert this

ericpgreen2 replied:
Yeah, I think we should avoid it there too :)

ericpgreen2 commented on web-admin/src/features/projects/status/ProjectResources.svelte:
```diff
@@ -2,73 +2,166 @@
+      }),
+      refetchInterval: (data) => {
+        if (
+          $resources?.isError ||
```
This still isn't a pure function, given `$resources.isError` is still here. I'm worried about a circular reference.

ericpgreen2 commented on web-admin/src/features/projects/status/ActionsCell.svelte:
```diff
@@ -0,0 +1,69 @@
+
+  const createTrigger = createRuntimeServiceCreateTrigger();
+
+  async function refresh(resourceKind: string, resourceName: string) {
+    // First trigger the refresh
+    await $createTrigger.mutateAsync({
+      instanceId: $runtime.instanceId,
+      data: {
+        resources: [
+          {
+            kind: resourceKind,
+            name: resourceName,
+          },
+        ],
+      },
+    });
+
+    // Then update UI state and invalidate query
+    triggerRefresh(resourceName);
+  }
```
It's a bit hard-to-follow that the refresh action has logic in 3 different places: this function, the inline function that starts on line 65, and the `triggerRefresh` callback function defined in `ProjectResources.svelte`. I'd propose unifying the functions in one place.

lovincyrus replied:
This was needed to pass the resourceName upstream to display all the notifications in ProjectResources. I'll unify the functions.

ericpgreen2 reviewed (changes requested):
When the user navigates away from the status page, we'll have to get rid of this toast message, else it'll spin forever.

<img width="1362" alt="image" src="https://github.com/user-attachments/assets/64a03bc7-dbcb-4700-9c7d-ff29af66f974" />

Tangentially, @ericokuma / @jkhwu, I'd propose we remove this toast message altogether, since it's redundant to the spinner in the table. And, in its current form, it's not accurately describing what's happening when: the initial resource has been refreshed successfully, and a downstream resource (or multiple) are refreshing.

<img width="1244" alt="image" src="https://github.com/user-attachments/assets/a1a49243-4c4a-4666-884f-ab964409d93a" />

lovincyrus commented on web-common/src/lib/event-bus/event-bus.ts:
```diff
@@ -52,6 +52,7 @@ export const eventBus = new EventBus();
 
 export interface Events {
   notification: NotificationMessage;
+  "clear-all-notifications": void;
```
I can remove these notification changes before landing @ericpgreen2

ericpgreen2 commented on web-admin/src/features/projects/status/ProjectResources.svelte:
```diff
@@ -2,54 +2,92 @@
+      }),
+      refetchInterval: (data) =>
+        calculateRefetchInterval(
+          $resources?.data ? currentRefetchInterval : INITIAL_REFETCH_INTERVAL,
```
The `currentRefetchInterval` isn't getting updated, so this is currently polling every 1s w/o a backoff.
